enhancements/update/targeted-update-edge-blocking.md

```diff
@@ -0,0 +1,966 @@
+
+Incoming bugs are evaluated to determine an impact statement based on [a generic template][rhbz-1858026-impact-statement-request].
+Some bugs only impact specific platforms, or clusters with other specific features.
+For example, rhbz#1858026 [only impacted][rhbz-1858026-impact-statement] clusters with the `None` platform which were created as 4.1 clusters and subsequently updated via 4.2, 4.3, and 4.4 to reach 4.5.
+And rhbz#1957584 [only impacted][rhbz-1957584-impact-statement] clusters updating from 4.6 to 4.7 with Routes whose `spec.host` contains no dots or was otherwise an invalid domain name.
+
+In those cases there is currently tension between wanting to protect vulnerable clusters by blocking the edge vs. wanting to avoid inconveniencing clusters which we know are not vulnerable and whose administrators may have been planning on taking the recommended update.
+This enhancement aims to reduce that tension.
+
+### Goals
+
+* [Cincinnati graph-data][graph-data] maintainers will have the ability to block edges for a vulnerable subset of clusters.
+
+### Non-Goals
+
+* Exactly scoping the set of blocked clusters to those which would have been impacted by the issue.
+    For example, some issues may be races where the impacted cluster set is a random subset of the vulnerable cluster set.
+    Any targeting of the blocked edges will reduce the number of blocked clusters which would have not been impacted, and thus reduce the contention between protecting vulnerable clusters and inconveniencing invulnerable clusters, so even overly broad scoping is better than no scoping at all.
+* Specifying a particular update service implementation.
+    This enhancement floats some ideas, but the details of the chosen approach are up to each update service's maintainers.
+
+## Proposal
+
+### Enhanced graph-data schema for blocking edges
+
+[The existing blocked-edges schema][block-edges] will be extended with the following new properties:
+
+* `url` (optional, [string][json-string]), with a URI documenting the blocking reason.
+    For example, this could link to a bug's impact statement or knowledge-base article.
+* `reason` (optional, [string][json-string]), with a CamelCase reason suitable for [a `ClusterOperatorStatusCondition` `reason` property][api-reason].
+* `message` (optional, [string][json-string]), with a human-oriented message describing the blocking reason, suitable for [a `ClusterOperatorStatusCondition` `message` property][api-message].
+* `matchingRules` (optional, [array][json-array]), defining conditions for deciding which clusters have the update recommended and which do not.
+  The array is ordered by decreasing precedence.
+  Consumers should walk the array in order, and stop after the first rule they can successfully evaluate.
+  If no condition can be successfully evaluated, the update should not be recommended.
```

Comment:
This description doesn't match what I remember from slack, either. I thought we were going for logical OR semantics? So, it's not "stop after the first rule they can successfully evaluate". It's "stop if a rule results in a True response" because at that point we know the cluster matches and we should not recommend the update, so we don't need to check any of the other rules.

Reply:
> if at least one rule evaluates to "True", is the edge recommended or not?
> if at least one rule evaluates to "False", is the edge recommended or not?
> ...

The reason we have precedence and exit early is because you could have different precision with different cluster-condition types.  For example, consider a bug on vSphere-based clusters with `http(s)Proxy` configured (manufacturing a complicated case).  We can express that exactly with the `promql` type, but with some false-positives with the (future, hypothetical) `platform` type:

```yaml
matchingRules:
- type: promql
  promql: |
    1 - group(cluster_infrastructure_provider{type=~"VSphere|None"}) * max(cluster_proxy_enabled{type=~"https?"})
    or
    group(cluster_infrastructure_provider) * group(cluster_proxy_enabled)
- type: platform
  sadPlatforms:
  - VSphere
  - None
```

Breaking down the possibilities:

* Cluster is vSphere with `httpsProxy`, CVO can evaluate the `promql` rule, `Recommended=False`.
* Cluster is vSphere with `httpsProxy`, CVO cannot evaluate the `promql` rule, but can evaluate the `platform` rule, `Recommended=False`.
* Cluster is vSphere with no Proxy config, CVO can evaluate the `promql` rule, `Recommended=True`.
* Cluster is vSphere with no Proxy config, CVO cannot evaluate the `promql` rule, but can evaluate the `platform` rule, `Recommended=False` (false-positive).

> It's "stop if a rule results in a True response" because at that point we know the cluster matches and we should not recommend the update...

If one rule is sad, that's sufficient to ignore later, less-precise rules.  But if one rule is happy, that's also sufficient to ignore later, less-precise rules.  Because if you had the same cluster state, and the CVO for some reason could not evaluate the the later rule type, it would be recommending the update.  So we cannot ship any rules that would give a false-negative and appear to evaluate happy despite the cluster actually being at risk.

Reply:
so:

1) walk down the rules in order
2) as soon as you hit a rule that evaluates to 1, block the edge, you're done.
3) as soon as you hit a rule that evaluates to 0, allow the edge, you're done.
4) as soon as you hit a rule you can't evaluate, block the edge, you're done.

is that it?

Reply:
> 0. if there are no rules defined, block the edge (this is the api compatibility case)

And also the "we're blocking this edge for everyone, but want to use `url`, `reason`, and `message` to explain why" case.  Where we currently use text-only erratas to get that information out.

> 2. if you hit a rule that evaluates to 0, keep processing rules

Nope, I'd just stop here, and say "rule happy, so `Recommended=True`".  And doesn't the *In which case, the cluster-version operator would* bit down in the *Non-PromQL filters* section make this clear?  Maybe the issue is my language here, where "can successfully evaluate" means "happy or sad" (not just "happy") and distinguishes from the "unknown/cannot-eval" case?

> 3. as soon as you hit a rule that evaluates to 1, block the edge, you're done.

Yes to this one.

> 4. as soon as you hit a rule you can't evaluate, block the edge, you're done.

Nope, if you can't evaluate, you continue to later rules, to see if you can evaluate those.

> 5. if you get to the end and didn't block the edge, allow the edge.

Nope.  If you have a happy result, you immediately `Recommended=True`, without attempting later rules.  If you have a sad result, you immediately `Recommended=False`, without attempting later rules.  If you hit an unknown result while walking the rule set, you are `Recommended=Unknown`.  If there are no rules in `matchingRules` at all, you are `Recommended=False`.

Reply:
I think the part that may be missing is possible values and what happens with the update recommendation based upon those values on line 75. This is confused because we do explain what the recommendation state is when all rules fail to evaluate on line 76.

Reply:
>>        as soon as you hit a rule that evaluates to 1, block the edge, you're done.
> Yes to this one.
>>       as soon as you hit a rule you can't evaluate, block the edge, you're done.
> Nope, if you can't evaluate, you continue to later rules, to see if you can evaluate those.

Maybe we have a different meaning for "can't evaluate", my meaning was "prom is down/we don't have data for some reason/we got a 503 on the request".  And if that is the meaning then the problem is that if an earlier rule *should have* blocked the edge, but instead we failed to evaluate it for some reason, and then a later rule allows the edge, you'll end up allowing the edge when you should have blocked it.

If you meant "can't evaluate" in the sense of "for some reason it doesn't make sense to evaluate this rule on this cluster", then i guess we need to list all the outcomes even more carefully.

it's also a little difficult to reason about what it means to have a rule evaluate as "success".  It implies that for the rule, the cluster matches the rule.  But since we ultimately want to "allow this edge if the cluster does NOT meet these conditions" you end up having to write the rules as "not has-condition-foo" so that when it evaluates to true the edge is blocked.

and yes, what scott said

Reply:
> If no condition can be successfully evaluated, the update should not be recommended.
>  If you hit an unknown result while walking the rule set, you are Recommended=Unknown

so if no condition can be evaluated, we're in recommended=false
but if you only hit one that can't be evaluated you're in recommended=unknown?

Fundamentally i think this idea of reasoning through ordered rules and having to consider which rule will actually become relevant depending on which results earlier rules get, is very complicated.

not clear why we wouldn't just AND rules together if/when we reach the point of needing multiple rules (even if one is a promql rule and another is a straight "platform=x" rule)


`!vsphere && !(baremetal+proxy) && !ppc64`

Reply:
`matchingRules` can probably be best understood by adding some concrete examples with explanations for each.

Reply:
How about:

>  Consumers should walk the array in order.  For a given entry, if a condition type is unrecognised, or fails to evaluate, consumers should proceed to the next entry.  If a condition successfully evaluates, that result is used, and no futher entries should be attempted.  If no condition can be successfully evaluated, the update should not be recommended.

Reply:
This is understandable to me. I thought we were going to use `match` but I'm not sure it's use would make this any more clear.

Reply:
matching vs. not matching are two subsets of successful evaluation.  How about:

> Consumers should walk the array in order. For a given entry, if a condition type is unrecognized, or fails to evaluate, consumers should proceed to the next entry. If a condition successfully evaluates (either as a match or as an explicit does-not-match), that result is used, and no further entries should be attempted. If no condition can be successfully evaluated, the update should not be recommended.

?

Reply:
@wking Lets go with your most recent suggestion in this thread. LGTM

Reply:
Pushed through with 1449253 -> fff9d3c.